### PR TITLE
chore: migrate `exec` behavior for maru

### PR DESCRIPTION
**Pull request by Racer159**
## Description

This migrates `exec` here in preparation of `maru` disconnecting itself from `zarf`.

## Related Issue

Fixes #N/A

### CODE REVIEW

File: exec/utils.go
Comment (Noxsios) [inline exec/utils.go]: package comment only needs to exist on 1 file to satisfy `revive`, I would keep it on `exec/exec.go` and remove from all others

File: exec/shell.go
Comment (Noxsios) [inline exec/shell.go]:
```suggestion
// IsPowerShell returns whether a shell name is PowerShell
func IsPowerShell(shellName string) bool {
```

i'm sorry, thats just what they named it

File: exec/utils.go
Comment (Noxsios) [inline exec/utils.go]:
```suggestion
	return filepath.EvalSymlinks(binaryPath)
```

File: exec/go.mod
Comment (Noxsios) [inline exec/go.mod]: this will need a local `replace` added (see `oci/go.mod`)

File: exec/exec.go
Comment (Noxsios) [inline exec/exec.go]: `PrintCfg` does not really need to exist, nor does `CmdWithPrint`

If a consumer wants to print, they can set `Print: true` on their `Config` struct.

There should only be 2 functions:

```
func Cmd(config Config, command string, args ...string) (string, string, error) {
	return CmdWithContext(context.TODO(), config, command, args...)
}

func CmdWithContext(ctx context.Context, config Config, command string, args ...string) (string, string, error) {
...
```

File: exec/exec.go
Comment (Noxsios) [inline exec/exec.go]: would it be cleaner to use an `errgroup` here?

File: exec/exec.go
Comment (Noxsios) [inline exec/exec.go:124]: this starts the command, and does not wait for it to be finished, does this not create a zombie process?

**Review by Noxsios — CHANGES_REQUESTED** [review]

File: exec/go.mod
Comment (Noxsios) [reply exec/go.mod]: all other packages will need the same treatment

File: exec/utils.go
Comment (Noxsios) [reply exec/utils.go]: all other packages will need the same treatment

File: exec/exec.go
Comment (Racer159) [reply exec/exec.go:124]: That is intentional so that when Zarf exits the browser doesn't.

File: exec/exec.go
Comment (AustinAbro321) [inline exec/exec.go:1]: We can probably reduce the duplication on these workflows, but we'll need workflows release-exec and release-variables created

File: variables/templates_test.go
Comment (AustinAbro321) [inline variables/templates_test.go]: This should be deferred so it gets deleted on failure

File: variables/templates_test.go
Comment (AustinAbro321) [inline variables/templates_test.go]: We should cleanup file.test automatically, also should be removed from the PR

**Review by AustinAbro321 — CHANGES_REQUESTED** [review]
Love the added unit tests! Nice work

File: variables/templates_test.go
Comment (Noxsios) [reply variables/templates_test.go]: If this file gets created in a `t.TempDir`, it will get cleaned up automatically, and will also prevent any accidental committing of test files

File: variables/templates_test.go
Comment (Noxsios) [reply variables/templates_test.go]: or run in a `t.Cleanup` (I would prefer the file get placed in a `t.TempDir` anyways)

File: variables/templates_test.go
Comment (Racer159) [reply variables/templates_test.go]: `file.test` never changes and felt it was better to have it in a file than in the test (since wrapper code would be needed to make / delete it).  I _can_ remove it though if you really want me to

File: variables/templates_test.go
Comment (Noxsios) [inline variables/templates_test.go]: this function is only used in 1 place, I would consider moving it there

File: helpers/progress.go
Comment (Noxsios) [inline helpers/progress.go:1]: changes in https://github.com/defenseunicorns/pkg/pull/85 should be removed from this PR

**Review by Noxsios — CHANGES_REQUESTED** [review]

File: variables/templates_test.go
Comment (Racer159) [reply variables/templates_test.go]: (also I intentionally committed it :sweat_smile: `git add variables/file.test -f` lol)

File: helpers/progress.go
Comment (Racer159) [reply helpers/progress.go:1]: once #85 is merged they will be smushed - I did it this way for continuity to https://github.com/defenseunicorns/maru-runner/pull/73

File: variables/templates_test.go
Comment (AustinAbro321) [reply variables/templates_test.go]: Ah I see it's an actual file variable. Initially I just assumed you accidentally added it since it was in the git ignore. Yeah I'm cool with keeping it

File: helpers/progress.go
Comment (Racer159) [reply helpers/progress.go:1]: 85 was smushed

File: variables/templates_test.go
Comment (AustinAbro321) [reply variables/templates_test.go]: I think we can help avoid confusion by adding the file into a directory called testdata
> Directory and file names that begin with "." or "_" are ignored by the go tool, as are directories named "testdata". 
https://pkg.go.dev/cmd/go#hdr-Package_lists_and_patterns

Comment (Racer159) [reply variables/templates_test.go]: perfect will do that

**Review by lucasrod16 — DISMISSED** [review]
lgtm

File: exec/exec.go
Comment (Noxsios) [inline exec/exec.go]: should it not be an `error` if you want to print, but have not given a `CommandPrinter`?

File: .gitignore
Comment (Noxsios) [inline .gitignore]:
```suggestion
```

Anything using this anymore?

File: exec/shell.go
Comment (Noxsios) [inline exec/shell.go]: this type should prob be `ShellPreference`, not `Shell`.  In the few cases it is used, it's assigned to a `shellPref` variable (which could be shortened to `pref`), and it gets confusing w/ the type being `Shell`, but there being variables passed around called `shell` that are not this type.

If there was a type `Shell` it would most likely be a `type Shell string`

File: exec/utils.go
Comment (Noxsios) [inline exec/utils.go]: Note that these functions are not go-routine safe as they edit a map w/ no protections?

Comment (Noxsios) [reply exec/utils.go]: Do we even want this to be a global registration system?

I think an exported map w/ methods might work better, there is no overlap between how Zarf does its `./zarf` mutation, UDS does its `./uds zarf` mutation, and Maru will do `./run` mutation?

**Review by Noxsios — CHANGES_REQUESTED** [review]

File: exec/exec.go
Comment (Racer159) [reply exec/exec.go]: There may be cases where you do want to print but not print the command itself (and have a more clean command output) - welcome thoughts on simplifying the interface though.

Comment (Racer159) [reply exec/exec.go:85]: (for example Maru doesn't give a command printer)

File: exec/utils.go
Comment (Racer159) [reply exec/utils.go]: This was not intended to be used in go-routines since it is most likely to be configured once within the root command or config code of an app - it was intended to be a way for multiple registrations to occur from different libraries - the thought was that `uds` could register where `./uds` was as well as `./zarf` (and maybe other shorthands like `./kubectl`).  We could move it to be passed into MutateCommand though there isn't an easy way to avoid globals between UDS and Maru because of how they are vendored (and the same would go for any `uds zarf` commands as well)

Comment (Racer159) [reply exec/utils.go]: switched to a sync map (though again I wouldn't recommend calling this function from goroutines because race conditions could result in weird outcomes and there shouldn't be a need to edit this map a ton in the first place).

File: exec/shell_test.go
Comment (Noxsios) [inline exec/shell_test.go]: rename symbols for `ShellPreference` type to be `shellPref` or `pref` to increase clarity

File: exec/utils_test.go
Comment (Noxsios) [inline exec/utils_test.go]: rename symbols for `ShellPreference` type to be `shellPref` or `pref` to increase clarity

**Review by Noxsios — CHANGES_REQUESTED** [review]

File: exec/exec.go
Comment (Noxsios) [reply exec/exec.go]: I think changing this to only `if config.CommandPrinter != nil ...`.

If a `CommandPrinter` has been given, that is explicit usage to print the command

Comment (Noxsios) [reply exec/exec.go]: And then you could print the command, but not print the output

File: exec/utils.go
Comment (Noxsios) [inline exec/utils.go:40]: unit test this function to ensure that:

- only strings in `supportedCmdMutation` are supported as keys, can be set properly, and return no error
- multiple calls overwrite
- an error is returned if an incorrect `cmdKey` is given

The only current place this function is called does not check this error, nor perform these checks

File: exec/utils.go
Comment (Noxsios) [inline exec/utils.go]:
```suggestion
// GetCmdMutation returns the cmdLocation for a given cmdKey and whether that key exists
func GetCmdMutation(cmdKey string) (string, bool) {
	if !slices.Contains(supportedCmdMutations, cmdKey) {
		return "", false
	}
	value, ok := registeredCmdMutations.Load(cmdKey)
	if !ok {
		return "", false
	}
	cmdLocation, ok := value.(string)
	if !ok {
		return "", false
	}
	return cmdLocation, true
}
```

Leave it up to the caller to decide what to do when the key they want does not exist.

This also re-orients the code to keep the golden path as least-indented as possible.

File: exec/utils_test.go
Comment (Noxsios) [inline exec/utils_test.go]: handle this error + unit test this guy

**Review by Noxsios — CHANGES_REQUESTED** [review]

File: exec/utils.go
Comment (Noxsios) [reply exec/utils.go:40]: Should this function prevent `""` from being set as a location?

Comment (Noxsios) [reply exec/utils.go:40]: Should this function check if a file exists at the given location and is executable?

File: exec/utils.go
Comment (Racer159) [reply exec/utils.go]: Hmm - wondering what the use case would be for not replacing?  If an app wanted to disable these they could register zarf to ./zarf and have it be turned off - having a default makes this easier to use for apps that don't have a location (since not all would).

File: exec/utils_test.go
Comment (Racer159) [reply exec/utils_test.go]: waiting to unit test until the API settles.

File: exec/utils.go
Comment (Racer159) [reply exec/utils.go:40]: waiting to unit test until the API settles.

**Review by AustinAbro321 — APPROVED** [review]
It seems strange to me that we're only allowing certain commands to be mutated, but I'm fine with that changing in a future PR if the need arises

**Review by Noxsios — APPROVED** [review]

**Pull request merged**
